sklearn/manifold/_barnes_hut_tsne.pyx

```diff
@@ -0,0 +1,682 @@
+# cython: boundscheck=False
+# cython: wraparound=False
+# cython: cdivision=True
+# Author: Christopher Moody <chrisemoody@gmail.com>
+# Author: Nick Travers <nickt@squareup.com>
+# Implementation by Chris Moody & Nick Travers
+# See http://homepage.tudelft.nl/19j49/t-SNE.html for reference
+# implementations and papers describing the technique
+
+
+from libc.stdlib cimport malloc, free
+from libc.stdio cimport printf
+from libc.math cimport sqrt
+cimport numpy as np
+import numpy as np
+
+cdef extern from "math.h":
+    float fabsf(float x) nogil
+
+# Round points differing by less than this amount
+# effectively ignoring differences near the 32bit 
+# floating point precision
+cdef float EPSILON = 1e-6
+
+# This is effectively an ifdef statement in Cython
+# It allows us to write printf debugging lines
+# and remove them at compile time
+cdef enum:
+    DEBUGFLAG = 0
+
+cdef extern from "time.h":
+    # Declare only what is necessary from `tm` structure.
+    ctypedef long clock_t
+    clock_t clock() nogil
+    double CLOCKS_PER_SEC
+
+
+cdef struct Node:
+    # Keep track of the center of mass
+    float[3] cum_com
+    # If this is a leaf, the position of the particle within this leaf 
+    float[3] cur_pos
+    # The number of particles including all 
+    # nodes below this one
+    long cum_size
+    # Number of particles at this node
+    long size
+    # Index of the particle at this node
+    long point_index
```

Comment:
This is defined only for non-empty leaf nodes, right? The comment could make that explicit.